src/backend/opencl/jit.cpp

```diff
@@ -66,90 +66,106 @@ static string getFuncName(Node *node, bool is_linear, bool *is_double)
     return hashName.str();
 }
 
-static string getKernelString(string funcName, Node *node, bool is_linear)
+static string getKernelString(string funcName, std::vector<Node *> nodes, bool is_linear)
 {
-    stringstream kerStream;
-    int id = node->getId();
 
-    kerStream << "__kernel void" << "\n";
+    // Common OpenCL code
+    // This part of the code does not change with the kernel.
+
+    static const char *kernelVoid =  "__kernel void\n";
+    static const char *dimParams = "KParam oInfo, uint groups_0, uint groups_1, uint num_odims";
+    static const char *blockStart = "{\n\n";
+    static const char *blockEnd = "\n\n}";
+
+    static const char *linearIndex = "\n"
+        "uint groupId  = get_group_id(1) * get_num_groups(0) + get_group_id(0);\n"
+        "uint threadId = get_local_id(0);\n"
+        "int idx = groupId * get_local_size(0) * get_local_size(1) + threadId;\n"
+        "if (idx >= oInfo.dims[3] * oInfo.strides[3]) return;\n";
+
+    static const char *generalIndex = "\n"
+        "uint id0 = 0, id1 = 0, id2 = 0, id3 = 0;\n"
+        "if (num_odims > 2) {\n"
+        "id2 = get_group_id(0) / groups_0;\n"
+        "id0 = get_group_id(0) - id2 * groups_0;\n"
+        "id0 = get_local_id(0) + id0 * get_local_size(0);\n"
+        "if (num_odims > 3) {\n"
+        "id3 = get_group_id(1) / groups_1;\n"
+        "id1 = get_group_id(1) - id3 * groups_1;\n"
+        "id1 = get_local_id(1) + id1 * get_local_size(1);\n"
+        "} else {\n"
+        "id1 = get_global_id(1);\n"
+        "}\n"
+        " } else {\n"
+        "id3 = 0;\n"
+        "id2 = 0;\n"
+        "id1 = get_global_id(1);\n"
+        "id0 = get_global_id(0);\n"
+        "}\n"
+        "bool cond = \n"
+        "id0 < oInfo.dims[0] && \n"
+        "id1 < oInfo.dims[1] && \n"
+        "id2 < oInfo.dims[2] && \n"
+        "id3 < oInfo.dims[3];\n\n"
+        "if (!cond) return;\n\n"
+        "int idx = "
+        "oInfo.strides[3] * id3 + oInfo.strides[2] * id2 + "
+        "oInfo.strides[1] * id1 + id0 + oInfo.offset;\n\n";
+
+
+    stringstream inParamStream;
+    stringstream outParamStream;
+    stringstream outWriteStream;
+    stringstream offsetsStream;
+    stringstream opsStream;
+
+    int count  = 0;
+
+    for (auto node : nodes) {
+        int id = node->getId();
+        node->genParams(inParamStream);
+        outParamStream << "__global " << node->getTypeStr() << " *out" << id << ", \n";
+        outWriteStream << "out" << id << "[idx] = " << "val" << id << ";\n";
+        node->genOffsets(offsetsStream, is_linear);
+        node->genFuncs(opsStream);
+        opsStream << "//" << ++count << std::endl << std::endl;
+    }
 
+    // Put various blocks into a single stream
+    stringstream kerStream;
+    kerStream << kernelVoid;
```

Comment:
I feel like many of these stream operations could be converted into string concatenations, but I don't know if it will be better for performance